core/src/replay_stage.rs

```diff
@@ -851,18 +950,51 @@ impl ReplayStage {
             descendants,
         );
 
-        for d in slot_descendants
-            .iter()
-            .chain(std::iter::once(&duplicate_slot))
-        {
-            // Clear the progress map of these forks
-            let _ = progress.remove(d);
+        // Grab the Slot and BankId's of the banks we need to purge, then clear the banks
+        // from BankForks
+        let (slots_to_purge, removed_banks): (Vec<(Slot, BankId)>, Vec<Arc<Bank>>) = {
+            let mut w_bank_forks = bank_forks.write().unwrap();
+            slot_descendants
+                .iter()
+                .chain(std::iter::once(&duplicate_slot))
+                .map(|slot| {
+                    // Clear the duplicate banks from BankForks
+                    let bank = w_bank_forks
+                        .remove(*slot)
+                        .expect("BankForks should not have been purged yet");
```

Comment:
Could we have a situation where the descendant was also a duplicate and purged first due to the ordering of `duplicate_slots_to_repair`?

Reply:
ha! refer to comment: https://github.com/solana-labs/solana/pull/18160#discussion_r665851025. 

I think this check https://github.com/solana-labs/solana/blob/b74ff2362aaf4fe4a14a0ce2cbc30f5c05dce33e/core/src/replay_stage.rs#L886-L890 should cover that case. Maybe it shouldn't be a warn?